staging/src/k8s.io/apiserver/pkg/util/proxy/streamtunnel.go

```diff
@@ -0,0 +1,454 @@
+		return
+	}
+
+	spdyRequest := createSPDYRequest(req, spdyProtocols...)
+
+	writer := &tunnelingResponseWriter{
+		w: w,
+		conn: &headerInterceptingConn{
+			initializableConn: &tunnelingWebsocketUpgraderConn{
+				w:   w,
+				req: req,
+			},
+		},
+	}
+
+	klog.V(4).Infoln("Tunnel spdy through websockets using the UpgradeAwareProxy")
+	h.upgradeHandler.ServeHTTP(writer, spdyRequest)
+}
+
+// createSPDYRequest modifies the passed request to remove
+// WebSockets headers and add SPDY upgrade information, including
+// spdy protocols acceptable to the client.
+func createSPDYRequest(req *http.Request, spdyProtocols ...string) *http.Request {
+	clone := utilnet.CloneRequest(req)
+	// Clean up the websocket headers from the http request.
+	clone.Header.Del(wsstream.WebSocketProtocolHeader)
+	clone.Header.Del("Sec-Websocket-Key")
+	clone.Header.Del("Sec-Websocket-Version")
+	clone.Header.Del(httpstream.HeaderUpgrade)
+	// Update the http request for an upstream SPDY upgrade.
+	clone.Method = "POST"
+	clone.Body = nil // Remove the request body which is unused.
+	clone.Header.Set(httpstream.HeaderUpgrade, spdy.HeaderSpdy31)
+	clone.Header.Del(httpstream.HeaderProtocolVersion)
+	for i := range spdyProtocols {
+		clone.Header.Add(httpstream.HeaderProtocolVersion, spdyProtocols[i])
+	}
+	return clone
+}
+
+// spdyProtocolsFromWebsocketProtocols returns a list of spdy protocols by filtering
+// to Kubernetes websocket subprotocols prefixed with "SPDY/3.1+", then removing the prefix
+func spdyProtocolsFromWebsocketProtocols(req *http.Request) []string {
+	var spdyProtocols []string
+	for _, protocol := range gwebsocket.Subprotocols(req) {
+		if strings.HasPrefix(protocol, constants.WebsocketsSPDYTunnelingPrefix) && strings.HasSuffix(protocol, constants.KubernetesSuffix) {
+			spdyProtocols = append(spdyProtocols, strings.TrimPrefix(protocol, constants.WebsocketsSPDYTunnelingPrefix))
+		}
+	}
+	return spdyProtocols
+}
+
+var _ http.ResponseWriter = &tunnelingResponseWriter{}
+var _ http.Hijacker = &tunnelingResponseWriter{}
+
+// tunnelingResponseWriter implements the http.ResponseWriter and http.Hijacker interfaces.
+// Only non-upgrade responses can be written using WriteHeader() and Write().
+// Once Write or WriteHeader is called, Hijack returns an error.
+// Once Hijack is called, Write, WriteHeader, and Hijack return errors.
+type tunnelingResponseWriter struct {
+	// w is used to delegate Header(), WriteHeader(), and Write() calls
+	w http.ResponseWriter
+	// conn is returned from Hijack()
+	conn net.Conn
+	// mu guards writes
+	mu sync.Mutex
+	// wrote tracks whether WriteHeader or Write has been called
+	written bool
+	// hijacked tracks whether Hijack has been called
+	hijacked bool
+}
+
+// Hijack returns a delegate "net.Conn".
+// An error is returned if Write(), WriteHeader(), or Hijack() was previously called.
+// The returned bufio.ReadWriter is always nil.
+func (w *tunnelingResponseWriter) Hijack() (net.Conn, *bufio.ReadWriter, error) {
```

Comment:
The std lib https://cs.opensource.google/go/go/+/faf882d1d427e8c8a9a1be00d8ddcab81d1e848e abort pending reads, before hijacking, I don't feel we have to, do we ?

Reply:
since in our use we don't do anything with the request, I don't think so